File: blockchain/src/main/java/pt/um/lei/masb/blockchain/Coinbase.java
package pt.um.lei.masb.blockchain;

import pt.um.lei.masb.blockchain.data.*;

import javax.persistence.*;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Positive;
import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.security.PublicKey;
import java.time.temporal.ChronoField;
import java.util.HashSet;
import java.util.Set;

/**
 * The coinbase transaction. Pays out to contributors to the blockchain.
 */
@Entity
public final class Coinbase implements Sizeable, IHashed {
    private static final int TIME_BASE = 5;
    private static final int VALUE_BASE = 2;
    private static final int BASE = 3;
    private static final int THRESHOLD = 100000;
    private static final int OTHER = 50;
    private static final int DATA = 5;
    private static final MathContext MATH_CONTEXT = new MathContext(8, RoundingMode.HALF_EVEN);


    @OneToMany(cascade = CascadeType.ALL,
               fetch = FetchType.EAGER,
               orphanRemoval = true)
    private final Set<TransactionOutput> payoutTXO;


    @Basic(optional = false)
    private BigDecimal coinbase;


    @Id
    private String hashId;

    /**
     * The coinbase will be continually updated
     * to reflect changes to the block.
     */
    protected Coinbase() {
        coinbase = new BigDecimal(0);
        payoutTXO = new HashSet<>();
    }

    public Coinbase(Set<TransactionOutput> payoutTXO,
                    BigDecimal coinbase,
                    String hashId) {
        this.payoutTXO = payoutTXO;
        this.coinbase = coinbase;
        this.hashId = hashId;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getHashId() {
        return hashId;
    }

    public @NotNull Set<TransactionOutput> getPayoutTXO() {
        return payoutTXO;
    }

    public @NotNull BigDecimal getCoinbase() {
        return coinbase;
    }

    /**
     * @param newT                  Transaction to contribute to payout.
     * @param latestKnown           Transaction to compare for fluctuation.
     * @param latestUTXO            Transaction with last unspent
     *                              transaction output for the new Transaction's publisher.
     *                              <p>
     *                              If it's the first time for this identity, supply the
     *                              origin block coinbase.
     * @param cat                   Category of the transaction's data.
     */
    protected void addToInput(@NotNull Transaction newT,
                              @NotNull Transaction latestKnown,
                              @NotNull Coinbase latestUTXO,
                              @NotNull Category cat) {
        var dt = newT.getSensorData();
        var dt2 = latestKnown.getSensorData();
        var deltaTime = getTimeDelta(dt, dt2);
        BigDecimal deltaValue;
        BigDecimal payout;
        switch (cat) {
            case TEMPERATURE:
                deltaValue = calculateDiffTemperature(dt.getTemperatureData(),
                                                      dt2.getTemperatureData());
                payout = calculateDiff(deltaTime, deltaValue, Coinbase.DATA);
                break;
            case LUMINOSITY:
                deltaValue = calculateDiffLuminosity(dt.getLuminosityData(),
                                                     dt2.getLuminosityData());
                payout = calculateDiff(deltaTime, deltaValue, Coinbase.DATA);
                break;
            case HUMIDITY:
                deltaValue = calculateDiffHumidity(dt.getHumidityData(),
                                                   dt2.getHumidityData());
                payout = calculateDiff(deltaTime, deltaValue, Coinbase.DATA);
                break;
            case NOISE:
                deltaValue = calculateDiffNoise(dt.getNoiseData(),
                                                dt2.getNoiseData());
                payout = calculateDiff(deltaTime, deltaValue, Coinbase.DATA);
                break;
            case OTHER:
                deltaValue = new BigDecimal(0);
                payout = calculateDiff(deltaTime, deltaValue, Coinbase.OTHER);
                break;
            default:
                payout = new BigDecimal(0);
        }
        coinbase = coinbase.add(payout);
        addToOutputs(newT.getPublicKey(), latestUTXO, payout);
    }


    private @NotNull BigDecimal calculateDiffTemperature(@NotNull TemperatureData newTD,
                                                         @NotNull TemperatureData oldTD) {
        var newT = newTD.convertToCelsius();
        var oldT = oldTD.convertToCelsius();
        return newT.subtract(oldT).divide(oldT, Coinbase.MATH_CONTEXT);
    }

    private @NotNull BigDecimal calculateDiffLuminosity(@NotNull LuminosityData newLD,
                                                        @NotNull LuminosityData oldLD) {
        return newLD.getLum()
                    .subtract(oldLD.getLum())
                    .divide(oldLD.getLum(), Coinbase.MATH_CONTEXT);
    }

    private @NotNull BigDecimal calculateDiffHumidity(@NotNull HumidityData newHD,
                                                      @NotNull HumidityData oldHD) {
        BigDecimal newH;
        BigDecimal oldH;
        if (newHD.getUnit() == HUnit.RELATIVE) {
            newH = newHD.getHum();
            oldH = oldHD.getHum();
        } else {
            newH = newHD.convertToKGbyKG();
            oldH = oldHD.convertToKGbyKG();
        }
        return newH.subtract(oldH).divide(oldH, Coinbase.MATH_CONTEXT);
    }

    private @NotNull BigDecimal calculateDiffNoise(@NotNull NoiseData newND,
                                                   @NotNull NoiseData oldND) {
        var newN = newND.getNoiseLevel().add(newND.getPeakOrBase()).abs();
        var oldN = oldND.getNoiseLevel().add(oldND.getPeakOrBase()).abs();
        return newN.subtract(oldN)
                   .divide(oldN, Coinbase.MATH_CONTEXT);
    }


    private @NotNull BigDecimal getTimeDelta(@NotNull SensorData dt,
                                             @NotNull SensorData dt2) {
        var stamp1 = new BigDecimal(dt.getTimestamp()
                                      .getEpochSecond() * 1000 + dt.getTimestamp()
                                                                   .get(ChronoField.MILLI_OF_SECOND));
        var stamp2 = new BigDecimal(dt2.getTimestamp()
                                       .getEpochSecond() * 1000 + dt2.getTimestamp()
                                                                     .get(ChronoField.MILLI_OF_SECOND));
        return stamp1.subtract(stamp2)
                     .divide(stamp1, new MathContext(8, RoundingMode.HALF_EVEN));
    }

    /**
     * @param publicKey Public Key of transaction publisher.
     * @param prevUTXO  Coinbase with previous known UTXO.
     * @param payout    Payout amount to publisher.
     */
    private void addToOutputs(@NotNull PublicKey publicKey,
                              @NotEmpty Coinbase prevUTXO,
                              @NotNull BigDecimal payout) {
        payoutTXO.stream()
                 .filter(t -> t.getPublicKey().equals(publicKey))
                 .findAny()
                 .ifPresentOrElse(t -> t.addToPayout(payout),
                                  () -> fillInFromPreviousUTXO(publicKey, prevUTXO, payout));
    }


    /**
     *
     * @param publicKey             The public identity associated with an agent.
     * @param prevUTXO              The coinbase containing the previous transaction
     *                              output due to the agent.
     * @param payout                The new payout to add to previous output.
     */
    private void fillInFromPreviousUTXO(@NotNull PublicKey publicKey,
                                        @NotNull Coinbase prevUTXO,
                                        @NotNull BigDecimal payout) {
        var newPayout = prevUTXO.payoutTXO.stream()
                                          .filter(t -> t.getPublicKey().equals(publicKey))
                                          .findAny()
                                          .map(TransactionOutput::getPayout)
                                          .orElseGet(() -> new BigDecimal(0)).add(payout);
        payoutTXO.add(new TransactionOutput(publicKey,
                                            prevUTXO.hashId,
                                            newPayout));
    }

    private @NotNull BigDecimal calculateDiff(@NotNull BigDecimal deltaTime,
                                              @NotNull BigDecimal deltaValue,
                                              @Positive int constant) {
        var standardDivisor = new BigDecimal(Coinbase.THRESHOLD * constant);
        var timeFactor = deltaTime.multiply(new BigDecimal(Coinbase.TIME_BASE))
                                  .pow(2, Coinbase.MATH_CONTEXT)
                                  .divide(standardDivisor, Coinbase.MATH_CONTEXT);
        var valueFactor = deltaValue.divide(new BigDecimal(2), Coinbase.MATH_CONTEXT)
                                    .multiply(new BigDecimal(Coinbase.VALUE_BASE))
                                    .divide(standardDivisor, Coinbase.MATH_CONTEXT);
        var baseFactor = new BigDecimal(Coinbase.BASE).divide(standardDivisor, Coinbase.MATH_CONTEXT);
        return timeFactor.add(valueFactor).add(baseFactor);
    }
}

File: ledger/src/main/kotlin/pt/um/masb/ledger/storage/Coinbase.kt
package pt.um.masb.ledger.storage

import com.squareup.moshi.JsonClass
import mu.KLogging
import org.openjdk.jol.info.ClassLayout
import pt.um.masb.common.Sizeable
import pt.um.masb.common.data.DataFormula
import pt.um.masb.common.data.Payout
import pt.um.masb.common.data.calculateDiff
import pt.um.masb.common.hash.AvailableHashAlgorithms
import pt.um.masb.common.hash.Hash
import pt.um.masb.common.hash.Hash.Companion.emptyHash
import pt.um.masb.common.hash.Hashable
import pt.um.masb.common.hash.Hashed
import pt.um.masb.common.hash.Hasher
import pt.um.masb.common.misc.flattenBytes
import pt.um.masb.common.storage.LedgerContract
import pt.um.masb.ledger.data.PhysicalData
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode
import java.security.PublicKey
import java.time.temporal.ChronoField

/**
 * The coinbase transaction. Pays out to contributors to the ledger.
 *
 * The coinbase will be continually updated
 * to reflect changes to the block.
 */
@JsonClass(generateAdapter = true)
data class Coinbase(
    val payoutTXO: MutableSet<TransactionOutput>,
    var coinbase: Payout,
    override var hashId: Hash,
    @Transient
    private val payoutFormula: DataFormula = ::calculateDiff
) : Sizeable, Hashed, Hashable, LedgerContract {


    override val approximateSize: Long
        get() {
            val sum = payoutTXO
                .fold(0) { acc: Long,
                           tOut: TransactionOutput ->
                    acc + tOut.approximateSize
                }
            return sum + ClassLayout
                .parseClass(this::class.java)
                .instanceSize()
        }

    init {
        if (hashId.contentEquals(
                emptyHash
            )
        ) {
            hashId = digest(crypter)
        }
    }

    constructor() : this(
        mutableSetOf(),
        Payout(BigDecimal.ZERO),
        emptyHash,
        ::calculateDiff
    )

    constructor(
        payoutTXO: MutableSet<TransactionOutput>,
        coinbase: Payout,
        hashId: Hash
    ) : this(
        payoutTXO,
        coinbase,
        hashId,
        ::calculateDiff
    )

    /**
     * Takes the new Transaction and attempts to calculate a fluctuation from
     * the previous Transaction of same type and in the same geographical area.
     *
     * Adds a payout for the transaction's agent in a transaction output.
     * @param newT                  Transaction to contribute to payout.
     * @param latestKnown           Transaction to compare for fluctuation.
     * @param latestUTXO            Last unspent transaction output for
     *                              the new Transaction's publisher.
     *                              <pw>
     *                              If it's the first time for this identity, supply
     *                              null.
     */
    internal fun addToInput(
        newT: Transaction,
        latestKnown: Transaction?,
        latestUTXO: TransactionOutput?
    ) {
        val payout: Payout
        val lkHash: Hash
        val lUTXOHash: Hash = latestUTXO?.hashId
            ?: emptyHash

        //None are known for this area.
        if (latestKnown == null) {
            payout = payoutFormula(
                BASE,
                TIME_BASE,
                BigDecimal.ONE,
                VALUE_BASE,
                BigDecimal.ONE,
                newT.data.dataConstant,
                THRESHOLD,
                MATH_CONTEXT
            )
            lkHash = emptyHash
        } else {
            payout = calculatePayout(
                newT.data,
                latestKnown.data
            )
            lkHash = latestKnown.hashId
        }
        coinbase = coinbase.add(payout)
        addToOutputs(
            newT.publicKey,
            lUTXOHash,
            newT.hashId,
            lkHash,
            payout
        )
        hashId = digest(crypter)
    }

    private fun getTimeDelta(
        dt: PhysicalData,
        dt2: PhysicalData
    ): BigDecimal {
        val stamp1 = BigDecimal(
            dt.instant.epochSecond * 1000000000 +
                    dt.instant.get(ChronoField.NANO_OF_SECOND)
        )
        val stamp2 = BigDecimal(
            dt2.instant.epochSecond * 1000000000 +
                    dt2.instant.get(ChronoField.NANO_OF_SECOND)
        )
        return stamp1
            .subtract(stamp2)
            .divide(stamp1, MATH_CONTEXT)
    }

    /**
     * @param publicKey Public Key of transaction publisher.
     * @param prevUTXO  Previous known UTXO's hashId.
     * @param newT      Transaction to contribute to payout's hashId.
     * @param prev      Transaction compared for fluctuation's hashId,
     *                  might be empty.
     * @param payout    Payout amount to publisher.
     */
    private fun addToOutputs(
        publicKey: PublicKey,
        prevUTXO: Hash,
        newT: Hash,
        prev: Hash,
        payout: Payout
    ) {
        payoutTXO
            .firstOrNull { it.publicKey == publicKey }
            .let {
                it?.addToPayout(
                    payout,
                    newT,
                    prev
                )
                    ?: payoutTXO.add(
                        TransactionOutput(
                            publicKey,
                            prevUTXO,
                            payout,
                            newT,
                            prev
                        )
                    )
            }
    }

    private fun calculatePayout(
        dt: PhysicalData,
        dt2: PhysicalData
    ): Payout =
        payoutFormula(
            BASE,
            TIME_BASE,
            getTimeDelta(dt, dt2),
            VALUE_BASE,
            dt.calculateDiff(dt2.data),
            dt.dataConstant,
            THRESHOLD,
            MATH_CONTEXT
        )

    override fun digest(c: Hasher): Hash =
        c.applyHash(
            flattenBytes(
                payoutTXO.sumBy {
                    it.hashId.bytes.size
                },
                payoutTXO.asSequence().map {
                    it.hashId.bytes
                },
                coinbase.bytes
            )
        )


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Coinbase) return false

        if (!payoutTXO.containsAll(other.payoutTXO)) return false
        if (payoutTXO.size != other.payoutTXO.size) return false
        if (coinbase != other.coinbase) return false
        if (!hashId.contentEquals(other.hashId)) return false
        if (payoutFormula != other.payoutFormula) return false

        return true
    }

    override fun hashCode(): Int {
        var result = payoutTXO.hashCode()
        result = 31 * result + coinbase.hashCode()
        result = 31 * result + hashId.contentHashCode()
        result = 31 * result + payoutFormula.hashCode()
        return result
    }

    companion object : KLogging() {
        const val TIME_BASE = 5
        const val VALUE_BASE = 2
        const val BASE = 3
        const val THRESHOLD = 100000
        const val OTHER = 50
        const val DATA = 5
        val MATH_CONTEXT = MathContext(
            12,
            RoundingMode.HALF_EVEN
        )
        val crypter = AvailableHashAlgorithms.SHA256Hasher
    }

}

